c10/core/impl/cow/state_machine.h

```diff
@@ -0,0 +1,104 @@
+  // was part of the original tensors that share a view.
+  //
+  // The generation of the output will come from:
+  // 1) maybe_shadow_storage->generation(), if non-null.
+  // 2) this->default_shadow_storage_->generation(), if it is set.
+  // 3) physical_generation_, i.e. 0, if this is the first lazy copy.
+  auto simulate_lazy_copy(cow::ShadowStorage* maybe_shadow_storage)
+      -> intrusive_ptr<cow::ShadowStorage>;
+
+ private:
+  // Identifies where a storage instance is in the lifecycle of
+  // tracking simulated lazy copies.
+  enum class StateId {
+    // This is the initial state. It may not be returned to and
+    // indicates that the storage has never had a lazy copy requested
+    // of it.
+    //
+    // INVARIANT: not default_shadow_storage_.has_value()
+    initial,
+
+    // This is transitioned to when the first copy on write takes
+    // place.
+    //
+    // INVARIANT: default_shadow_storage_.has_value()
+    active,
+
+    // TODO Consider adding an "inactive" state. An inactive state
+    //      would occur when the last outstanding lazy copy goes away
+    //      or has been materialized. At that point, we may stop
+    //      tracking generations.
+    //
+    //      We may want to do this because we've already warned or
+    //      because we only had a temporary lazy copy that shouldn't
+    //      affect the performance of the program for the remainder of
+    //      its lifetime.
+    //
+    //      DANGER! There could be outstanding shadow storages. For
+    //      example, if the last set of tensors sharing a view
+    //      remaining was not using the default shadow storage, it
+    //      will continue holding the instances that they share.
+    //
+    //      There are a few potential solutions to this. One is for
+    //      the state machine to be the exclusive owner of the shadow
+    //      storages and for tensors to only hold weak pointers to
+    //      them.
+    //
+    //      Another is for tensors to hold a storage id and an index
+    //      and those can be invalidated by increasing the storage id
+    //      whenever we transition to StateId::inactive. A nice
+    //      property of this is that it would generalize the design
+    //      decision around the "default shadow storage".
+  };
+
+  // The current state of the storage instance.
+  std::atomic<StateId> state_id_;
+
+  // Guards all the state.
+  std::mutex mtx_;
+  // How many writes have been applied to the storage.
+  std::uint64_t physical_generation_;
+  // The shadow storage to use for any tensors that don't have
+  // one. This situation is common, and will be true for tensors and
+  // views thereof created before any copy on writes.
+  //
+  // Note: this would like to be std::optional, but it can't be
+  // because of torchdeploy incompatibility.
+  //
+  // See https://github.com/pytorch/multipy/issues/312
+  c10::optional<cow::ShadowStorageNonIntrusive> default_shadow_storage_;
```

Comment:
This is a lot of fields. Mutex is like 80 bytes, and then you also have four more words worth of stuff. Making matters worse, this struct is included inline in StorageImpl, which means we're paying this cost on every storage allocation. This will definitely affect Caffe2 perf.

Reply:
We talked over VC about whether or not it was actually necessary mutex protect these fields. My intuition is that it is not necessary. Additionally, use of c10::optional is costing you another word on default_shadow_storage_, but in fact what's the problem with just always having a ShadowStorageNonIntrusive and using state_id to control whether or not it is set or not?

Reply:
I added the ability to disable this on the TensorImpl side with the `PYTORCH_INSTRUMENT_COW_TENSOR` preprocessor definition. I implemented this to take up no storage on TensorImpl if it's not defined. Presumably we want the same here?

I tested this earlier, std::mutex is 40 bytes in both GCC and Clang.

We can cut down on the size a little. We can make the state_id_ one byte (assuming atomic supports that). We can then also make our generations 32-bit. And we can drop the `c10::optional` around the `default_shadow_storage_` and make its presence be contingent on the value of `state_id_`.

All together we should be at:
40 bytes: mutex
4 bytes: generation
4 bytes: default shadow_storage
1 byte: state id

The alignment of mutex is 8, so we don't gain anything by decreasing state id's size.

Is there a lighter weight mutex we can use?

Alternatively, I assume you are suggesting that we heap allocate the state machine in StorageImpl on demand, so we only pay the weight when there is a lazy copy active. We can easily do that.

Reply:
OK, this is in place. StorageImpl only pays a pointer size when this is not in use. I didn't implement the empty-base class optimization for it yet, but I can add that too, it's really easy.

Reply:
OK, so let's discuss whether or not we still need a Mutex here. Let's assume it doesn't exist.

In bump, we atomically increase the physical generation then, in a separate atomic operation, we atomically update the shadow storage of the tensor being written to.

Tensors that share a shadow storage are tensors that are views of one another, so they may not be mutated concurrently. However, tensors who share a "real" storage are logically lazy copies, and they may be mutated concurrently.

So if a concurrent write is taking place to two lazy copies, in the future world, at worst, both will materialize their own copies, and there's no issue.

In the legacy world, these are both physically views and they will conflict. We would like a warning to take place. The atomic bumps to the physical generation will not race, so one of them will see N + 1 and the other will see N + 2.

We should be able to safely just atomically set the shadow generations for each and one of them will warn.



Now let's see what will happen with a read that is racing with a write. Again, this is forbidden if these share a shadow storage, because a user is required to synchronize writes and reads to logically aliasing tensors.

So let's have a write to one lazy copy and a read from another racing.

Read thread:
R1) atomically read physical generation (shared with write)
R2) atomically read shadow generation (distinct from write)

Write thread:
W1) atomically increment physical generation (shared with read)
W2) atomically update shadow generation (distinct from read)

The order of R2 and W2 don't matter, so let's ignore those.

W1 -> R1: The reader sees the updated physical generation and will warn correctly.
R1 -> W1: The reader sees the old physical generation and will not warn. This may or may not be a false negative, depending on how the actual write and reads are sequenced. This is, however, sketchy in the old regime.

Nevertheless, the important thing is that there are no false positives. So I think we can safely drop the mutex. What do you think?

Reply:
#98504 for an idea of what i mean